### PR TITLE
Add default system sets based on `World` accesses

### LINKED ISSUE #4872: Automatically create labels for events?
## What problem does this solve or what need does it fill?

I've found that I keep creating labels for systems that write events so that I can create a fence between the readers and writers. I've also done this with ResMut systems.

## What solution would you like?

I don't know if this is a good idea, but it might be helpful to auto generate labels for events and resources (maybe even mutable queries if it's cheap) that get applied to systems that write to those things. Then you can add systems that go before or after anything that writes to an event, resource, or component.

## What alternative(s) have you considered?

Manually label stuff which is error prone. Then again it's not that hard and I don't know if the automation is worth it.

### LINKED ISSUE #7857: Add systems to sets based on their world accesses
## What problem does this solve or what need does it fill?

Currently, it is difficult to order systems based only on their logical constraints. In cases where you want a system to run after (or before) other systems which access a resource or component, you need to have global knowledge of the schedule and which systems might access that data.

One example of this is for events: if you have a system that reads `AppExit` events, it must run *after* the system that sends those events -- otherwise the event will be missed forever. Ordering the systems correctly requires knowing which system sends the event and adding a dependency to it, which is difficult to discover. See #7067, which is motivated by this issue. 

## What solution would you like?

Make it possible to automatically assign systems to a set based on their world accesses. An event reader system with proper ordering could look like:

```rust
fn read_app_exit(mut events: EventReader<AppExit>) { ... }

app.add_system(read_app_exit.after(SendsEvents::<AppExit>::default());
```

When combined with #7838, transform propagation could be as simple as:

```rust
#[derive(Component)]
#[component(write_set = WriteTransform)] // any system with access to `&mut Transform` will be added to this set.
pub struct Transform { ... }

// Perform transform propagation after every system has modified `Transform`,
// and after their commands have been flushed.
app.add_system(transform_propagate_system.after_and_flush(WriteTransform));
```

There should be a way of manually opting out of access sets:

```rust
fn jump(q: Query<&mut Transform>) { ... }

// This system can run after `transform_propagate_system`.
app.add_system(jump.without_set(WritesTransform));
```

## What alternative(s) have you considered?

Do nothing, and continue using global reasoning to order systems.

## Additional context

A variant of this feature that is restricted to events was originally suggested in #4872.

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

Resolve #7857.

Give users more tools to order systems based on logical constraints. Allow automatically grouping systems with specific world accesses or system params into sets, which can be used to specify dependencies.

## Solution

Any component, resource, or event can optionally be associated with a `SystemSet`, which will be assigned to any system that accesses that world data.

Example: TODO

---
## Changelog

todo

## Migration Guide

todo

### CODE REVIEW

**Comment by JoJoJet** [comment]
> Why not define this for all data accesses, which would then encapsulate events?

This pattern could also be useful for `GlobalTransform` propagation, I think. We could mark the system with `.after(MutSystem::<Transform>::default())`

I definitely agree with that direction, but I wanted feedback from a few members before doing the implementation work.

**Comment by JoJoJet** [comment]
Also, I think that if we make auto-labels ubiquitous, we'll have to consider letting users explicitly opt out of specific auto-labels for their systems.

**Comment by alice-i-cecile** [comment]
> Also, I think that if we make auto-labels ubiquitous, we'll have to consider letting users explicitly opt out of specific auto-labels for their systems.

Agreed, I think this is essential even here.

**Comment by JoJoJet** [comment]
> Bikeshed: I think I prefer `WritesTo` / `ReadFrom` as names. These should be clear from context that they're system labels.

Even better: make the type name the verb, and the method name the preposition. `Writes::to::<MyEvent>()`. We'd probably need a different preposition other than `from`, though.

File: crates/bevy_ecs/src/schedule/system_descriptor.rs [hunk 1]
```diff
@@ -126,6 +126,9 @@ pub trait ParallelSystemDescriptorCoercion<Params> {
     /// Assigns a label to the system; there can be more than one, and it doesn't have to be unique.
     fn label(self, label: impl SystemLabel) -> ParallelSystemDescriptor;
 
+    /// Removes the specified label from this system. Does nothing if the label is not present.
+    fn ignoring(self, label: impl SystemLabel) -> ParallelSystemDescriptor;
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/system_descriptor.rs:130]: Please bikeshed

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/schedule/system_descriptor.rs:130]: `unlabel`? `remove_label`?

**Comment by alice-i-cecile** [comment]
My first impression is that I'd be fine with that. The weak benchmarks here definitely make me nervous though.

We could also store two HashSets, one for "labels" and the other for "labels this system definitely doesn't have", and then check reactively.

**Comment by JoJoJet** [comment]
It appears that labels already use a `HashMap` for looking up dependencies, so having extra labels shouldn't impact performance too severely.

Using the fancy new benchmarks at #5410, I added 50 distinct, unused labels to every single system.

| # of plugins | time | % change |
|-|-|-|
| 10 | 1.3344 ms | +10.700% |
| 50 | 41.205 ms | +2.5983% |
| 100 | 186.23 ms | +0.4773% |

10% regression isn't bad, considering that's 10\*20\*50=10,000 unused labels. We can also see that the performance impact becomes negligible as you add more and more systems.

**Comment by JoJoJet** [comment]
Thoughts on expanding this to components
* Should be off by default and discouraged
  * Adding dozens of unused labels to every system is gross, even if the benchmarks are fine.
  * Deferred mutations via commands are ubiquitous and invisible to this API, so extra care needs to be taken to ensure everything is correct.
* We should consider `Transform`/`GlobalTransform` to be the "canonical" use-case, so we need to make sure this API meets those needs.
  * The end-`Update` command flush must come after anything marked with `Writes<Transform>` or `FlushTransform`.
  * `transform_propagate_system` must come after the end-`Update` command flush.
  * Any system containing `Commands` gets auto-labeled with `FlushTransform` (strawman)
    * Users may explicitly remove the `FlushTransform` label if they want systems with commands to come after `transform_propagate_system`.
    * This is similar to current proposals of automatically placing systems before `transform_propogate_system` by default. Except here, the dependency is only applied when there's actually a chance that a system might mutate `Transform`.
  * `transform_propogate_system` is set to run before any system with `Reads<GlobalTransform>`
    * As always, users may explicitly remove the `Reads<GlobalTransform>` label if they wish.

**Comment by alice-i-cecile** [comment]
Yeah, I'm wondering if we can configure the opt-in and opt-out behavior via metadata on the Component / Resource / Event traits.

And yes, another point in favor of trying to make commands more transparent.

The other thought is that this PR will become dramatically more useful once cross-stage ordering dependencies work. We may want to fix that before merging this.

**Comment by JoJoJet** [comment]
> Yeah, I'm wondering if we can configure the opt-in and opt-out behavior via metadata on the Component / Resource / Event traits.

We should definitely do this for components. Having to derive every resource would be annoying IMO.  
It's probably unnecessary for events, so we shouldn't force users to derive them just for this one thing.

**Comment by alice-i-cecile** [comment]
This would close #7857, correct?

**Comment by JoJoJet** [comment]
If we decide to go with the macro-based approach, then yes :)

**Comment by JoJoJet** [comment]
Closing this out, since I don't think the derive macro-based approach is the right way to handle this.

**Pull request closed without merging**